docker_test.go

```diff
@@ -146,270 +88,566 @@ type Event struct {
 	ip                string
 }
 
-func createWorkspaceConfig(templatePath string, values map[string]string) string {
-	t, err := template.ParseFiles(templatePath)
-	if err != nil {
-		panic(err)
+func TestMainFlow(t *testing.T) {
+	runSlow = config.GetEnvAsBool("SLOW", false)
+	if !runSlow {
+		t.Skip("Skipping tests. Use 'SLOW=1' env var to run them.")
 	}
 
-	f, err := os.CreateTemp("", "workspaceConfig.*.json")
-	if err != nil {
-		panic(err)
+	hold = config.GetEnvAsBool("HOLD", false)
+	runBigQueryTest = config.GetEnvAsBool("BIGQUERYINTEGRATION", false)
+	if os.Getenv("OVERRIDE_ARM64_CHECK") == "1" {
+		overrideArm64Check = true
 	}
-	defer func() { _ = f.Close() }()
 
-	err = t.Execute(f, values)
-	if err != nil {
-		panic(err)
-	}
+	var tearDownStart time.Time
+	defer func() {
+		if tearDownStart == (time.Time{}) {
+			t.Log("--- Teardown done (unexpected)")
+		} else {
+			t.Logf("--- Teardown done (%s)", time.Since(tearDownStart))
+		}
+	}()
 
-	return f.Name()
-}
+	svcCtx, svcCancel := context.WithCancel(context.Background())
+	svcDone := setupMainFlow(svcCtx, t)
+	sendEventsToGateway(t)
 
-func waitUntilReady(ctx context.Context, endpoint string, atMost, interval time.Duration, caller string) {
-	probe := time.NewTicker(interval)
-	timeout := time.After(atMost)
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-timeout:
-			log.Panicf("application was not ready after %s, for the end point: %s, caller: %s\n", atMost, endpoint,
-				caller)
-		case <-probe.C:
-			resp, err := http.Get(endpoint)
-			if err != nil {
-				continue
+	t.Run("webhook", func(t *testing.T) {
```

Comment:
for some of these sub-tests we could even turn on the parallel flag

Reply:
Yes although there is still no perfect isolation, I'd like to achieve that one first. Despite decoupling the sending of the events I'm still not able to have the "kafka" test pass alone for example. once I'm able to achieve that I'll try to turn the parallelism on.

One additional note on parallelism is that I had that on my kafka tests and I was able to run them in less than 30s on my desktop, but it was making them extremely flaky on CI so I had to remove it. It seems that with the 2 cores of the default CI env, too much parallelism might cause timeouts. So even though I'm not against it might require a bit of tuning to get it right.

Reply:
We have started containers once, even if the tests are still coupled, i.e. operate on the same source, at least using parallelism we can have them all fail/pass faster.